File: web/src/utils/avatar.ts
const ALLOWED_MIME = new Set(["image/png", "image/jpeg", "image/webp"]);
export const MAX_AVATAR_BYTES = 256 * 1024;

export function isTrustedAvatarDataUrl(value?: string | null): boolean {
  if (!value) return false;
  return /^data:image\/(png|jpeg|jpg|webp);base64,[a-zA-Z0-9+/=]+$/.test(value);
}

export function validateAvatarFile(file: File): string | null {
  if (!ALLOWED_MIME.has(file.type)) {
    return "Formato inválido. Use PNG, JPEG ou WEBP.";
  }
  if (file.size <= 0 || file.size > MAX_AVATAR_BYTES) {
    return "Avatar muito grande. Limite de 256KB.";
  }
  return null;
}

export function hasAllowedMagicBytes(buffer: ArrayBuffer): boolean {
  const b = new Uint8Array(buffer);
  const isPNG =
    b.length >= 8 &&
    b[0] === 0x89 &&
    b[1] === 0x50 &&
    b[2] === 0x4e &&
    b[3] === 0x47 &&
    b[4] === 0x0d &&
    b[5] === 0x0a &&
    b[6] === 0x1a &&
    b[7] === 0x0a;
  const isJPEG = b.length >= 3 && b[0] === 0xff && b[1] === 0xd8 && b[2] === 0xff;
  const isWEBP =
    b.length >= 12 &&
    b[0] === 0x52 &&
    b[1] === 0x49 &&
    b[2] === 0x46 &&
    b[3] === 0x46 &&
    b[8] === 0x57 &&
    b[9] === 0x45 &&
    b[10] === 0x42 &&
    b[11] === 0x50;
  return isPNG || isJPEG || isWEBP;
}

File: server/internal/utils/avatar.go
package utils

import (
	"encoding/base64"
	"errors"
	"strings"
)

const (
	MaxAvatarBytes = 256 * 1024
)

// ValidateAvatarDataURL accepts only inline data URLs for png/jpeg/webp.
// This prevents external image trackers and validates payload by magic bytes.
func ValidateAvatarDataURL(v string) error {
	if v == "" {
		return nil
	}
	if len(v) > 500000 {
		return errors.New("avatar too large")
	}

	prefixes := []string{
		"data:image/png;base64,",
		"data:image/jpeg;base64,",
		"data:image/jpg;base64,",
		"data:image/webp;base64,",
	}

	var payload string
	validPrefix := false
	for _, p := range prefixes {
		if strings.HasPrefix(v, p) {
			payload = strings.TrimPrefix(v, p)
			validPrefix = true
			break
		}
	}
	if !validPrefix {
		return errors.New("invalid avatar format")
	}

	decoded, err := base64.StdEncoding.DecodeString(payload)
	if err != nil {
		return errors.New("invalid avatar base64")
	}
	if len(decoded) == 0 || len(decoded) > MaxAvatarBytes {
		return errors.New("avatar size is invalid")
	}
	if !hasAllowedMagicBytes(decoded) {
		return errors.New("unsupported avatar type")
	}

	return nil
}

func hasAllowedMagicBytes(b []byte) bool {
	// PNG: 89 50 4E 47 0D 0A 1A 0A
	if len(b) >= 8 &&
		b[0] == 0x89 && b[1] == 0x50 && b[2] == 0x4E && b[3] == 0x47 &&
		b[4] == 0x0D && b[5] == 0x0A && b[6] == 0x1A && b[7] == 0x0A {
		return true
	}
	// JPEG: FF D8 FF
	if len(b) >= 3 && b[0] == 0xFF && b[1] == 0xD8 && b[2] == 0xFF {
		return true
	}
	// WEBP: RIFF....WEBP
	if len(b) >= 12 &&
		b[0] == 'R' && b[1] == 'I' && b[2] == 'F' && b[3] == 'F' &&
		b[8] == 'W' && b[9] == 'E' && b[10] == 'B' && b[11] == 'P' {
		return true
	}
	return false
}

